### PR TITLE
sqld programs

### PR DESCRIPTION
**Pull request by MarinPostma**
This PR implements sqld programs on the model of hrana programs.

Programs take over batches in the execution of queries within sqld, since they are very flexible and let the `Database` API consumer specify how statements should be executed.

The API of the sqld programs closely matches that of hrana batches, so the implementation on hrana side is changed to simply transform a hrana batch into a sqld program.

On the HTTP side, batches of queries are transformed into a program that causes a rollback if any statement in the batch fails.

On the replica, the programs are serialized and sent over by gRPC to the primary when we detect that they could cause a write. 

This PR also solves a known bug in HTTP batches that used to unconditionally execute statements in a batch even if a statement in the batch failed, and never rolledback, since the new implementation of HTTP batches ensures that a rollback is always performed if any statement in the batch fails to execute.

Future work: expose a more direct access to the `Database` to the consumers

### CODE REVIEW

File: sqld/src/database/write_proxy.rs [hunk 1]
```diff
@@ -118,39 +119,33 @@ impl WriteProxyDatabase {
...
 #[async_trait::async_trait]
 impl Database for WriteProxyDatabase {
-    async fn execute_batch(&self, queries: query::Queries) -> Result<(Vec<QueryResult>, State)> {
+    async fn execute_program(&self, pgm: Program) -> Result<(Vec<Option<QueryResult>>, State)> {
         let mut state = self.state.lock().await;
         if *state == State::Init
-            && queries.iter().all(|q| q.stmt.is_read_only())
-            && final_state(*state, queries.iter().map(|s| &s.stmt)) == State::Init
+            && pgm.is_read_only()
+            && final_state(*state, pgm.steps.iter().map(|s| &s.query.stmt)) == State::Init
```
Comment (honzasp) [inline sqld/src/database/write_proxy.rs:126]: I'm not sure if this is correct: some statements in the program can be skipped, but `final_state()` assumes that we always execute all statements.

File: sqld/src/hrana/batch.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+};
+
+#[derive(thiserror::Error, Debug)]
+pub enum BatchError {
+    #[error("Invalid reference to step in a condition")]
+    CondBadStep,
+}
+
+fn proto_cond_to_cond(cond: &proto::BatchCond) -> Result<Cond> {
+    let try_convert_step = |step: i32| -> std::result::Result<usize, BatchError> {
```
Comment (honzasp) [inline sqld/src/hrana/batch.rs:18]: When we are already checking the step index, perhaps we can also check that it refers to step that is before the current step?

File: sqld/src/query_analysis.rs [hunk 3]
```diff
@@ -70,7 +70,11 @@ pub enum State {
 impl State {
     pub fn step(&mut self, kind: StmtKind) {
         *self = match (*self, kind) {
-            (State::Txn, StmtKind::TxnBegin) | (State::Init, StmtKind::TxnEnd) => State::Invalid,
+            // those two transition will cause an error, but since we are interested in what the
+            // pesimistic final state is, and we will adjust when we get the actual state back.
+            (State::Txn, StmtKind::TxnBegin) => State::Txn,
+            (State::Init, StmtKind::TxnEnd) => State::Init,
```
Comment (honzasp) [inline sqld/src/query_analysis.rs:76]: What does it mean that a state is pesimistic? Why not use `State::Invalid` as the result of these invalid transitions?

File: sqld/src/database/libsql.rs [hunk 4]
```diff
@@ -266,11 +185,153 @@ impl LibSqlDb {
...
+    ) -> Option<QueryResult> {
+        let enabled = match step.cond.as_ref() {
+            Some(cond) => match eval_cond(cond, results) {
+                Ok(enabled) => enabled,
+                Err(e) => return Some(Err(e)),
+            },
+            None => true,
+        };
+
+        enabled.then(|| self.execute_query(&step.query))
```
Comment (honzasp) [inline sqld/src/database/libsql.rs:231]: Nice! I didn't know about `bool::then()`, but I often thought that it would be useful :)

**Review by honzasp — COMMENTED** [review]
Looks ok, just a few comments (one important about correctness).

File: sqld/src/database/mod.rs [hunk 5]
```diff
@@ -9,17 +9,88 @@ pub mod write_proxy;
...
-    /// Executes a batch of queries, and returns a vec of results corresponding to the queries,
-    /// and the state the database is in after the call to execute.
-    async fn execute_batch(&self, queries: Queries) -> Result<(Vec<QueryResult>, State)>;
+    /// Execute all the queries in the batch sequentially.
+    /// If an query in the batch fails, the remaining queries are ignores, and the batch current
+    /// transaction (if any) is rolledback.
+    async fn execute_batch_or_rollback(
+        &self,
+        batch: Vec<Query>,
+    ) -> Result<(Vec<Option<QueryResult>>, State)> {
```
[the remark this answers is not in the document]
Comment (MarinPostma) [reply sqld/src/database/mod.rs:59]: there may or may not be an ongoing transaction. If there isn't then this is no_op, the rollback will fail and we ignore it, it there is an ongoing transaction, then we rollback

File: sqld/src/hrana/batch.rs (hunk 2 above)
Comment (MarinPostma) [reply sqld/src/hrana/batch.rs:18]: this is done when the program is executed there: https://github.com/libsql/sqld/pull/307/files#diff-b7c90481e4da77d124d59b67dd5c2141bfba58deac7df43808defee65815fef9R308

File: sqld/src/database/write_proxy.rs (hunk 1 above)
Comment (MarinPostma) [reply sqld/src/database/write_proxy.rs]: That's right, I went a bit too fast over this. I pushed a commit that address this issue, lmk

File: sqld/src/database/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (honzasp) [reply sqld/src/database/mod.rs:69]: Yeah, I understand what the code is doing, but the reason for this behavior is not clear to me? (I'm not saying it's a problem, I would just like to understand :))

File: sqld/src/database/write_proxy.rs [hunk 6]
```diff
@@ -115,50 +115,67 @@ impl WriteProxyDatabase {
...
-                    // Set state to invalid, so next call is sent to remote, and we have a chance
-                    // to recover state.
-                    *state = State::Invalid;
-                    Err(Error::RpcQueryExecutionError(e))
+        if *state == State::Init && pgm.is_read_only() {
+            // This program won't perform any write, we try to predict it's final state, i.e, if
+            // it's an interactive transaction or not, and then attempt to execute it. If it turns
+            // out that it was indeed an interactive transaction, we rollback the current
+            // transaction, and attempt to execute again on the remote
+            if predict_final_state(*state, pgm.steps.iter().map(|s| &s.query.stmt)) == State::Init {
```
Comment (honzasp) [inline sqld/src/database/write_proxy.rs:166]: I think that this check will always fail for batches that are produced by libsql client. The batch looks like `BEGIN`, ..., `COMMIT`, `ROLLBACK`, so the `predict_final_state()` function will always return `State::Invalid`, right?

Comment (honzasp) [reply sqld/src/database/write_proxy.rs:166]: If we verify that all statements are read-only, then perhaps we can always just execute the batch locally, and if this leaves a transaction open, then rollback it and retry on master? In effect, do the same as now, without checking for the predicted final state.

File: sqld/src/database/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MarinPostma) [reply sqld/src/database/mod.rs:69]: It is up to the client to decide whether the content of the batch should be executed from within a transaction, the only concern of executing a batch from the HTTP perspective is that it doesn't leave a transaction open after all statements have been executed

**Comment by MarinPostma** [comment]
bors merge

**Pull request merged**